File: Easy/589. N-ary Tree Preorder Traversal/iterative-solution.js
var preorder = function (root) {
    const list = [];
    const stack = [];

    let current = root;
    stack.push(current);

    while (stack.length > 0 && current) {
        current = stack.pop();
        list.push(current.val);

        for (let i = current.children.length - 1; i >= 0; i--) {
            stack.push(current.children[i]);
        }
    }

    return list;
};

File: 1. Easy/589. N-ary Tree Preorder Traversal/IterativeSolution.java
class Solution {
    public List<Integer> preorder(Node root) {
        Stack<Node> stack = new Stack<>();
        List<Integer> list = new ArrayList<>();

        Node current = root;
        stack.push(root);

        while (!stack.isEmpty() && current != null) {
            current = stack.pop();
            list.add(current.val);

            Collections.reverse(current.children);
            for (Node child : current.children) {
                stack.push(child);
            }
        }

        return list;
    }
}